### PR TITLE
Avoid exception when db-connection is lost

### LINKED ISSUE #707: Undefined Method Error Caused By Active?
Using version 5.1.6, activerecord 5.1.7

I tracked down the error to these lines. Starting from here:
```
def active?
        return false unless @connection
        raw_connection_do 'SELECT 1'
        true
      rescue *connection_errors
        false
      end
```
Raw connection is then called:
```
def raw_connection_do(sql)
          case @connection_options[:mode]
          when :dblib
            @connection.execute(sql).do
          end
        ensure
          @update_sql = false
        end
```

I believe the error is caused because `@connection.execute(sql)` can potentially return false. This causes the `undefined method do for false class error`. But I have not confirmed it.

Stacktrace:
```
NoMethodError: undefined method `do' for false:FalseClass
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.1.6/lib/active_record/connection_adapters/sqlserver/database_statements.rb:284:in `raw_connection_do'
/usr/local/bundle/gems/activerecord-sqlserver-adapter-5.1.6/lib/active_record/connection_adapters/sqlserver_adapter.rb:155:in `active?'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:444:in `verify!'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:810:in `block in checkout_and_verify'
/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:131:in `run_callbacks'
/usr/local/bundle/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:827:in `_run_checkout_callbacks'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:809:in `checkout_and_verify'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:502:in `checkout'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:376:in `connection'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:933:in `retrieve_connection'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_handling.rb:116:in `retrieve_connection'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/connection_handling.rb:88:in `connection'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/core.rb:301:in `cached_find_by_statement'
/usr/local/bundle/gems/activerecord-5.1.7/lib/active_record/core.rb:183:in `find'
```

### PR DESCRIPTION
**Pull request by Svelix**
The tiny_tds lib might return false on execute when the call fails.
This caused exceptions when active? is called after the connection
was lost.

fixes #707

### CODE REVIEW

File: lib/active_record/connection_adapters/sqlserver/database_statements.rb [hunk 1]
```diff
@@ -281,7 +281,8 @@ def sp_executesql_sql(sql, types, params, name)
         def raw_connection_do(sql)
           case @connection_options[:mode]
           when :dblib
-            @connection.execute(sql).do
+            result = @connection.execute(sql)
+            result == false ? false : result.do
```
Comment (scbabacus-midnight) [inline lib/active_record/connection_adapters/sqlserver/database_statements.rb]: I know this is essentially the same thing, but
~~~ruby
result && result.do
~~~
might be more concise.

I did not review the functionality of the change, though.  
I am just telling you an idiomatic ruby trick.

File: lib/active_record/connection_adapters/sqlserver_adapter.rb [hunk 2]
```diff
@@ -152,8 +152,7 @@ def disable_referential_integrity
 
       def active?
         return false unless @connection
-        raw_connection_do 'SELECT 1'
-        true
+        raw_connection_do('SELECT 1') == 1
```
Comment (scbabacus-midnight) [inline lib/active_record/connection_adapters/sqlserver_adapter.rb]: IMO the previous one before the change is more robust.
Because the return value is independent of `raw_connection_do` implementation.

No matter how `raw_connection_do` changes in the future, as long as still raise expected exceptions.
This function will still work correctly.

File: lib/active_record/connection_adapters/sqlserver/database_statements.rb (hunk 1 above)
Comment (Svelix) [reply lib/active_record/connection_adapters/sqlserver/database_statements.rb]: No, it's not the same.
```
result && result.do
```
will return result if it is any falsy value.
When you look at the tiny_tds lib, you will see that it should either returns false or a result object which then can be called do on it.
In any other case - for example `nil` - I don't want it to return nil, but cause an exception, so it's easier to figure out what is going wrong.

File: lib/active_record/connection_adapters/sqlserver_adapter.rb (hunk 2 above)
Comment (Svelix) [reply lib/active_record/connection_adapters/sqlserver_adapter.rb]: The problem with the previous version is that if raw_connection do does not return 1, but for example `false`, like it currently does when the connection is gone, no exception is raised which caused `active?` to return true, although the connection is not working.
With this change it will now only return true if the connection actually works, and false if an exception is raised, or anything else causes the connection to not return proper values any more.

**Comment by midnight-wonderer** [comment]
Could you elaborate more why or how `@connection.execute(sql)` returns false in the first place?
I am quite interested in this case now.

**Comment by Svelix** [comment]
> Could you elaborate more why or how `@connection.execute(sql)` returns false in the first place? Or how to reproduce the issue.
> 
> I am quite interested in this case now. This can potential be a bug of the underlying connection layer.

Its source is in the c-implementaion of the connection adapter: https://github.com/rails-sqlserver/tiny_tds/blob/master/ext/tiny_tds/client.c#L256

The simplest way I found to reproduce this locally is by using a docker-container for the database, open a Rails console and make any db-request, so a connection is opened. Then restart the database container without closing the Rails console and after the database docker container is up again, call the `active?` method.

File: lib/active_record/connection_adapters/sqlserver_adapter.rb (hunk 2 above)
Comment (Svelix) [reply lib/active_record/connection_adapters/sqlserver_adapter.rb]: Additionally I'd like to point out that I did not get rid of the rescue block, so the previous behaviour of returning false if an connection_errors is raised is still the same.

**Comment by scbabacus-midnight** [comment]
I opened an issue on TinyTds. (https://github.com/rails-sqlserver/tiny_tds/issues/451)

In the meantime, I found out that `#active?` is part of TinyTds public API.  
It used to have a delegation there https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/commit/a347f556a312bb4d56a9c8ebfcc2f28cc135641b
I wonder why it has been removed.

@metaskills 
Can we use `respond_to?` here instead, and delegate `active?` check to TinyTds?
"ActiveRecord and the connection pool" clearly does not work as intended because of the TinyTds itself.

To summarize the options now are:
- A) merge this PR
- B) raise a proper error from TinyTds when the execution failed
- C) delegate `active?` to TinyTds

B) and C) are not mutually exclusive, both can be done.

**Comment by Svelix** [comment]
From my experience `active?` from TinyTds does not return the desired status, but something else. Haven't looked into in in depth, but I remember having tried to use that instead and it didn't work out for me.

**Comment by wpolicarpo** [comment]
Hi. I'm jumping into this issue now.

I would prefer to have this fixed at `tiny_tds` level than the adapter level, but I'm not against this change as a temporary fix for now.

My only questions are

1) Should we really return `false` in `raw_connection_do` if something goes wrong with database connection? Shouldn't we raise an error instead just like `tiny_tds` would ideally do?
2) If so, what type of exception should be raised? `ActiveRecord::ConnectionTimeoutError`? `TinyTds::Error`?

@Svelix do you mind rebasing your branch so we can check for any failing tests?

**Comment by wpolicarpo** [comment]
Closing this via #818, #825 and rails-sqlserver/tiny_tds/pull/469.
